Realtime data implementation

CmpCtrl opened the pull request:
First off, I am new to python and opensource development and I have no idea what I don't know, so help me out if I'm missing anything and don't assume I know what I'm doing...

Anyway, I've been working on implementing the MQTT over websocket connection for streaming the real-time power usage data. That seems to be 1 msg per 3 seconds and contains a timestamp, the current power usage, and some accumulating counter, presumably power consumption. I think I have it basically functional, but I'm sure there's a lot of room for improvement. I guess I'm not sure how complete it should be before submitting a pull request, but it seems like a good stage to at least get some feedback. 

I had to update a handful of functions in the client.py to gather the relevant authentication info for the MQTT connection, and then add another request to start the fast polling on the smart meter. Then, I added a new script, realtime.py that handles the MQTT connection. That makes use of the paho-mqtt client library, [paho-mqtt](https://www.eclipse.org/paho/). This is done in a new class, `DukeEnergyRealtime`, which takes a `DukeEnergyClient` object as an input. I plan to subclass it in order to customize the `on_msg()` callback method to put the data in a database or whatever. The main method of this class is `connect_and_subscribe()` which will connect to the mqtt server and subscribe to the topic and run forever, as messages come in the `on_msg()` callback will be called. Ctrl-c can break out of this forever loop and will disconnect from the mqtt server first. There is a realtimeExample.py script that shows this. 

I came across a handful of issues while working on this. First, I started out working in a dev container, but something about that seemed to be setting the TCP_KEEPIDLE to 0 seconds so it would start sending TCP keepalive messages immediately. I'm not sure that's a real issue, but it seemed like a lot of unnecessary traffic and I struggled to adjust that. I ended up going to a virtual environment in vscode for my development which got around that issue. However, that highlighted that in windows, python 3.8+ switches the default loop for asyncio, so `asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())` is required before the `asyncio.run(main())` command.

Setting the SSLKEYLOGFILE environment variable to a writable text file is handy for decrypting the traffic with wireshark for instance (you can tell it where the keylogfile is in edit -> preferences -> protocol -> TLS -> (Pre)-Master-Secret log... Also, this wireshark plugin was very helpful [wireshark-mqtt-ws-plugin](https://github.com/nberthet/wireshark-mqtt-ws-plugin)

There are a handful of things I'm not fully happy with yet. First, when escaping from the loop it still throws a traceback even tho it does unsubscribe and disconnect as I intend. I'm not sure how to avoid that, or if it matters.

I don't yet have much error handling, nor do I know how long you can stream data for. I suspect it will eventually timeout, and I expect that the start fast poll request will need to be re-sent at some point. I think that the paho-mqtt library would try to reconnect, but that wouldn't request the fast polling again (due to the paho-mqtt callbacks not being asynchronous I cant await the fastpoll request there).

I'm not yet sure how well using subclassing to customize the on_msg callback will work in practice. While most of the paho-mqtt library is asynchronous, the callbacks are blocking. I think it will be OK in this particular case since the data rate is 1/3Hz but it doesn't feel like the right place to deal with the data. I'd love some suggestions on how best to handle that. 

My focus has been on something that could run continually to push the data into a db, so I haven't given much thought to other use-cases.

Anyway, let me know your feedback/thoughts/suggestions. 
Thanks,
Phil

mjmeli commented:
Hey, this is awesome! Thanks for digging into the real time data. I got stumped with authentication and I wonder if I just forgot to call the fastpoll/start endpoint. 

I'll find some time to review this ASAP

CmpCtrl commented:
Cool sounds good. Yea it will let you subscribe to a topic ok, but doesn't seem to rx any messages until you request the fast poll. Maybe it would rx very low rate messages. 

I did run into an issue this afternoon where it seemed to hang after running for a few minutes. It got stuck after sending a PINGREQ and receiving a PINGRES. That seems like an issue with the asyncio stuff which I'm really not comfortable with.

I missed including the paho-mqtt library in the setup.cfg, so that's been updated with some linting updates. I haven't had much luck with the testing tho, I don't know much about that so I'll try to learn a bit about it next week.

mjmeli commented on src/pyduke_energy/client.py:
```diff
@@ -177,6 +182,35 @@ async def get_gateway_usage(
+        try:
+            headers = await self._get_gateway_auth_headers()
+        except InputError:
+            # Assume 1st meter in 1st account # if missing
+            _LOGGER.info("No meter specified, assuming fist meter of first accnt")
+            accounts = await self.get_account_list()
+            meters = await self.get_account_details(accounts[0])
+            self.select_meter(meters.meter_infos[0])
+            headers = await self._get_gateway_auth_headers()
```
From my experience when I integrated this into Home Assistant, this isn't going to be very reliable. Most of the issues I've dealt with have been people with different account setups, e.g. multiple accounts, accounts with dumb meters, etc.

I'd also prefer that these functions keep failing if the proper meter isn't selected.

But I do like the idea of integrating code that finds the 1st meter automatically. What if we moved this into a separate function, like `select_default_meter()`, and then dropped the try/except from here. We could then move the logic I've built out for HA into it.

CmpCtrl replied:
Yea that makes sense, I'm happy to do that. I'll get back to working on this stuff Monday.

mjmeli commented on src/pyduke_energy/realtime.py:
```diff
@@ -0,0 +1,244 @@
+class DukeEnergyRealtime:
+    """Duke Energy Realtime Client."""
+
+    endpoint = "/app-mqtt"
```
can this be put in the const file?

CmpCtrl replied:
Certainly. I had noticed that most of the endpoints in the client.py were not in the constants, so I followed that convention, but yes, seems like they should be.

mjmeli replied:
good point...oops :)

mjmeli commented on src/pyduke_energy/realtime.py:
```diff
@@ -0,0 +1,244 @@
+            _LOGGER.warning("Unexpected message: %s", msg)
+        else:
+            self.rx_msg.set_result((msg.payload.decode("utf8")))
+            self.on_msg(msg)
```
maybe check on_msg is not None here in case no callback was implemented. since that seems like a weird case, you could raise an Exception

CmpCtrl replied:
My intention was for there to be a "default" on_msg callback in the class, in this case it just prints to the logger. But then if you want to subclass the DukeEnergyRealtime ( I think thats the right term) and define a more specialized version of on_msg it would allow you to do so. Does that make sense? is that a good way to accomplish this? It still doesn't feel ideal to me since the on_msg callback is blocking. Again, I'm very new to python, so i might be missing something obvious.

mjmeli replied:
I think the approach is good. So what I'm pointing out is that in theory, someone could just not implement `on_msg` or even use the `DukeEnergyRealtime` class without subclassing it. If you want this to be allowed (it sounds like you do), then you will have to consider the case that self.on_msg doesn't exist, which means the line here will throw an error.

Something like this will handle it

```
if self.on_msg:
    self.on_msg(msg)
```

If you don't want this to be a possibility, then you'd be best served marking the `DukeEnergyRealtime` and the method `on_msg` as abstract (https://stackoverflow.com/a/13646263).

mjmeli commented:
So overall I think it's pretty good. I pointed out a couple of things above to look at. 

It does also look like the linter is a little nitpicky. FYI if you want to run that locally you can do `tox -e lint`, since I apparently have to approve its runs here

Other more general things. Sometimes when I run, I get this happen where it doesn't receive any messages and eventually runs out of memory. Any idea what's going on here?

    INFO:pyduke_energy.client:No meter specified, assuming fist meter of first accnt
    DEBUG:pyduke_energy.client:Getting new OAuth auth
    DEBUG:pyduke_energy.client:Getting new gateway auth
    DEBUG:pyduke_energy.realtime:Socket opened
    DEBUG:pyduke_energy.realtime:Sending CONNECT (u1, p1, wr0, wq0, wf0, c1, k60) client_id=b'mobileSub1-P210400905-755f0d4d-bc73-4162-8b43-5027c6d10f7a'
    DEBUG:pyduke_energy.realtime:Misc loop started
    DEBUG:pyduke_energy.client:Smartmeter fastpoll requested
    DEBUG:pyduke_energy.realtime:Socket readable, calling loop_read()
    DEBUG:pyduke_energy.realtime:Received CONNACK (0, 0)
    DEBUG:pyduke_energy.realtime:MQTT connected with result code: Connection Accepted.
    DEBUG:pyduke_energy.realtime:Sending SUBSCRIBE (d0, m1) [(b'DESH/P210400905/out/sm/1/live', 0)]
    DEBUG:pyduke_energy.realtime:Watching socket for writability.
    DEBUG:pyduke_energy.realtime:Socket is writable, calling loop_write
    DEBUG:pyduke_energy.realtime:Stop watching socket for writability.
    DEBUG:pyduke_energy.realtime:Socket readable, calling loop_read()
    DEBUG:pyduke_energy.realtime:Received SUBACK
    DEBUG:pyduke_energy.realtime:MQTT subscribed msg_id: 1 qos: (0,)
    DEBUG:pyduke_energy.realtime:Socket readable, calling loop_read()
    DEBUG:pyduke_energy.realtime:Socket closed
    ERROR:pyduke_energy.realtime:MQTT disconnect error, result code: Out of memory.
    DEBUG:pyduke_energy.realtime:Misc loop finished

>Setting the SSLKEYLOGFILE environment variable to a writable text file is handy for decrypting the traffic with wireshark for instance

Can you explain this some more? Do I create a text file and set SSLKEYLOGFILE={path}?

> First, when escaping from the loop it still throws a traceback even tho it does unsubscribe and disconnect as I intend. I'm not sure how to avoid that, or if it matters.

I'm guessing this is because the KeyboardInterrupt is happening asynchronously. I'm not nearly an expert on Python so I don't really know what's happening, but I believe your `except KeyboardInterrupt:` in `DukeEnergyRealtime.connect_and_subscribe` isn't catching it, even though the `finally` runs. You can put that `except` around the `asyncio.run(main())` if you want in your example. I don't really understand it, but I agree it might not matter as the unsubscribe and disconnect is running. Maybe just drop that except block that you have now as it isn't working.

>I'm not yet sure how well using subclassing to customize the on_msg callback will work in practice. While most of the paho-mqtt library is asynchronous, the callbacks are blocking. I think it will be OK in this particular case since the data rate is 1/3Hz but it doesn't feel like the right place to deal with the data. I'd love some suggestions on how best to handle that.

I'm guessing the best practice is going to have to be to have the `on_msg` callback spin up another thread for the work to run on. That could be up to the consumer or maybe that could be built into the `_on_msg` where instead of calling `on_msg` directly, it puts it onto a thread.

CmpCtrl commented:
Thanks for following up on this quickly. I'll review it in more detail monday, but a couple quick answers below. 

I havent had much success with the testing or linting locally, i cant get it to run. i'll have to diagnose that more.

Yea, i had a similar issue once when it had run for several minutes after rx'ing a PINGRES msg. I think it must be something wrong in the asyncio loop where it doesnt trigger the _on_msg callback or maybe the loop_read, then the socket closes when it runs out of memory ( there are still messages coming in). i must not be awaiting properly somewhere. 

The SSLKEYLOGFILE environment variable can be set to a path of a writable text doc somewhere (not sure if it needs to exist already or not) but then it will dump the pre-master keys into there. I think most browsers honor that environment variable, and python does as well, if you use the `ssl.create_default_context()`. Then you can tell wireshark where that is and it will decrypt the tcp traffic. See the pre-master-secret here [https://wiki.wireshark.org/TLS](https://wiki.wireshark.org/TLS)

I thought the try->except->finally was needed to get the finally section to run, but i'll review that.

Thanks,
Phil

CmpCtrl commented:
Ok, here are a round of updates.

I think I figured out my linting and testing issues. hopefully those are better this time. 

I moved a variety of constants into the const.py, from both the realtime.py and client.py. 

I separated the logic to select the default meter into a new method. Feel free to tune that up, I just have the naive first meter of first account logic. 

Help me understand your concern with sub classing and the `on_msg()` method. I think that the `on_msg()` method should always exist since it is in the base class. Maybe its confusing in that there is an `_on_msg()` as well as an `on_msg()`. I did the private method to be sure that the `rx_msg.set_result()` would be handled appropriately despite potentially having the `on_msg()` method be re-defined in a subclass. Maybe I am misunderstanding how sub classing works, I don't mind adding the check for if its missing, but I just want to make sure I understand why its needed.

It seems like the fast polling times out after 15 mins, so i added a crude timer to call that again periodically. 

I still dont have a good handle on asyncio, but i think that the `KeyboardInterrupt `is caught at the `asyncio.run(main())` line. That explains why the `try: finally:` works to disconnect, but we still see the traceback. I'm not sure how best to address this yet.

mjmeli commented on src/pyduke_energy/client.py:
```diff
@@ -182,6 +187,12 @@ async def get_gateway_usage(
         measurements.sort(key=lambda x: x.timestamp)
         return measurements
 
+    async def select_default_meter(self):
```
This is good for now, I can add my more complex logic outside of this PR

mjmeli commented on src/pyduke_energy/const.py:
```diff
@@ -3,7 +3,18 @@
 CUST_API_BASE_URL = "https://cust-api.duke-energy.com/gep/v2/"
 CUST_PILOT_API_BASE_URL = "https://cust-pilot-api.duke-energy.com/"
 IOT_API_BASE_URL = "https://app-core1.de-iot.io/rest/cloud/"
+FASTPOLL_ENDPOINT = "smartmeter/fastpoll/start"
```
thx for doing these

mjmeli commented on src/pyduke_energy/realtime.py:
```diff
@@ -161,23 +170,26 @@ async def connect_and_subscribe(self):
+                    time.perf_counter() - tstart > FASTPOLL_TIMEOUT
+                ):  # Call fastpoll again
+                    tstart = await self.duke_energy.start_smartmeter_fastpoll()
                 await asyncio.sleep(1)
```
just curious (I missed this before) but why is this sleep necessary? couldn't this just move ahead to the await?

mjmeli commented:
Thanks for making those changes!

I see your confusion now with the `on_msg` stuff. I missed that there was an `on_msg` on the base class too. So yeah, no changes are needed there.

On the asyncio/KeyboardInterrupt thing, I'm not sure either. Honestly I'd just remove the except in the client and let the error bubble up. You don't need the except for the finally to run on exception (the finally will always run).

I'm still seeing the issue with the out of memory but maybe you are still looking in that?

Looks like the linting is good now so that's good.

I'm fine with iterative PRs so let me know when you want me to merge this in. The releases are cut independently of the merges so nothing actually gets released until I cut the release.

mjmeli commented:
I saw the memory issue once earlier but I haven't since. I'll keep trying and see if I can pull the full debug info.

mjmeli commented:
Alright, sorry again for the delay. I'm going to merge this now and then do some housekeeping steps on the repo to tick the version, update the README, and push to PyPi. Unless I screw something up, it will be v0.0.6.
